In [71]:
import numpy as np
np.set_printoptions(legacy='1.25')
import tensorflow as tf

In [43]:
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()

### Preprocessing, encoding etc.

In [59]:
# After doing the data processing, one hot encoding etc , build a CNN model and measure its accuracy and try to optimize the model

x_train_norm = (x_train).reshape((60000,28,28,1)) / 255 # reshaping and normalizing the data
x_test_norm = (x_test).reshape((10000,28,28,1)) / 255

y_train_enc = tf.keras.utils.to_categorical(y_train) # one hot encoding the targets
y_test_enc = tf.keras.utils.to_categorical(y_test)

### Making the model

In [65]:
model = tf.keras.models.Sequential()
model.add(tf.keras.layers.Conv2D(16, (3,3), activation='relu', input_shape=(28,28,1), padding='same'))
model.add(tf.keras.layers.MaxPool2D(pool_size=(2,2), strides=2, padding='valid'))
model.add(tf.keras.layers.Conv2D(8, (3,3), activation='relu', padding='same'))
model.add(tf.keras.layers.Flatten())
model.add(tf.keras.layers.Dense(10, activation='softmax'))
model.summary()

c:\Program Files\Python313\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential_5"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d_4 (Conv2D)               │ (None, 28, 28, 16)     │           160 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d (MaxPooling2D)    │ (None, 14, 14, 16)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_5 (Conv2D)               │ (None, 14, 14, 8)      │         1,160 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten_2 (Flatten)             │ (None, 1568)           │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 10)             │        15,690 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 17,010 (66.45 KB)

 Trainable params: 17,010 (66.45 KB)

 Non-trainable params: 0 (0.00 B)

In [66]:
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

In [69]:
model.fit(x_train_norm, y_train_enc, validation_split=0.15, batch_size=10, epochs=10)

Epoch 1/10
5100/5100 ━━━━━━━━━━━━━━━━━━━━ 8s 2ms/step - accuracy: 0.9958 - loss: 0.0129 - val_accuracy: 0.9872 - val_loss: 0.0584
Epoch 2/10
5100/5100 ━━━━━━━━━━━━━━━━━━━━ 8s 2ms/step - accuracy: 0.9962 - loss: 0.0110 - val_accuracy: 0.9871 - val_loss: 0.0613
Epoch 3/10
5100/5100 ━━━━━━━━━━━━━━━━━━━━ 8s 2ms/step - accuracy: 0.9964 - loss: 0.0100 - val_accuracy: 0.9854 - val_loss: 0.0854
Epoch 4/10
5100/5100 ━━━━━━━━━━━━━━━━━━━━ 8s 2ms/step - accuracy: 0.9968 - loss: 0.0092 - val_accuracy: 0.9857 - val_loss: 0.0716
Epoch 5/10
5100/5100 ━━━━━━━━━━━━━━━━━━━━ 8s 2ms/step - accuracy: 0.9969 - loss: 0.0083 - val_accuracy: 0.9880 - val_loss: 0.0713
Epoch 6/10
5100/5100 ━━━━━━━━━━━━━━━━━━━━ 8s 2ms/step - accuracy: 0.9972 - loss: 0.0083 - val_accuracy: 0.9858 - val_loss: 0.0832
Epoch 7/10
5100/5100 ━━━━━━━━━━━━━━━━━━━━ 8s 2ms/step - accuracy: 0.9976 - loss: 0.0070 - val_accuracy: 0.9883 - val_loss: 0.0691
Epoch 8/10
5100/5100 ━━━━━━━━━━━━━━━━━━━━ 8s 2ms/step - accuracy: 0.9979 - loss: 0.0057 - 

In [70]:
pred_data = model.predict(x_test_norm)
correct_pred_index = np.argmax(pred_data, axis=1)
correct_test_index = np.argmax(y_test_enc, axis=1)
accuracy = np.mean(correct_pred_index == correct_test_index) * 100
accuracy

313/313 ━━━━━━━━━━━━━━━━━━━━ 0s 873us/step


98.66

## Performance:

f: Flatten layer <br/>
p: Max Polling layer <br />
OF: Overfit
| Hidden Layers  | Epochs  | Accuracy    | Time   |
| -------------- | ------- | ----------- | ------ |
| 2 (32,f)       | 10      | 97.92%      | 1m 49s |
| 4 (16,p,8,f)   | 10      | 98.71%      | 1m 25s |
| 4 (16,p,8,f)   | 20      | 98.66%      | 2m 47s |